### PR TITLE
Fix ProfileTailCallNaked so it doesn't write xmm1 past the end of its stack

### PR DESCRIPTION
**Pull request by davmason**
Also do some documentation and refactoring.

### CODE REVIEW

File: src/coreclr/src/vm/amd64/asmhelpers.S [hunk 1]
```diff
@@ -1,6 +1,6 @@
-// Licensed to the .NET Foundation under one or more agreements.
-// The .NET Foundation licenses this file to you under the MIT license.
-// See the LICENSE file in the project root for more information.
+# Licensed to the .NET Foundation under one or more agreements.
```
Comment (lpereira) [inline src/coreclr/src/vm/amd64/asmhelpers.S]: Why the change in comment style?  Makes it difficult to see the actual thing this PR is fixing.

Comment (davmason) [reply src/coreclr/src/vm/amd64/asmhelpers.S]: The C style comments were messing with my editors code highlighting and since github will highlight the changed part of the line and this change is very small I figured it wouldn't impact readability that much.

File: src/coreclr/src/vm/amd64/asmhelpers.S [hunk 2]
```diff
@@ -9,141 +9,180 @@
...
 
-.equ SIZEOF_STACK_FRAME, SIZEOF_PROFILE_PLATFORM_SPECIFIC_DATA + SIZEOF_FP_ARG_SPILL
+.equ OFFSETOF_FP_ARG_SPILL, SIZEOF_PROFILE_PLATFORM_SPECIFIC_DATA
+
+# SIZEOF_STACK_FRAME is how many bytes we reserve in our ELT helpers below
+# There are three components, the first is space for profiler platform specific
+# data struct that we spill the general purpose registers to, then space to
+# spill xmm0 and xmm1, then finally 8 bytes of padding to ensure that the xmm
+# register reads/writes are aligned on 16 bytes.
+.equ SIZEOF_STACK_FRAME, SIZEOF_PROFILE_PLATFORM_SPECIFIC_DATA + SIZEOF_FP_ARG_SPILL + STACK_FUDGE_FACTOR
```
Comment (noahfalk) [inline src/coreclr/src/vm/amd64/asmhelpers.S]: defined above

File: src/coreclr/src/vm/amd64/asmhelpers.S [hunk 3]
```diff
@@ -9,141 +9,180 @@
...
+#
+#   Register preservation scheme:
+#
+#       Preserved:
+#           - all non-volatile registers
+#           - rax, rdx
+#           - xmm0, xmm1
+#
+#       Not Preserved:
+#           - integer argument registers (rcx, rdx, r8, r9)
```
Comment (noahfalk) [inline src/coreclr/src/vm/amd64/asmhelpers.S]: It predates your change but may as well fix it with other inconsistencies, the code does appear to preserve rcx, rdx, r8, and r9. The comment is also self-inconsistent claiming above that rdx was preserved + claiming it is not preserved here.

File: src/coreclr/src/vm/amd64/asmhelpers.S [hunk 4]
```diff
@@ -9,141 +9,180 @@
...
+.equ SIZEOF_PROFILE_PLATFORM_SPECIFIC_DATA, 0x8*22 + SIZEOF_PROFILE_STRUCT_BUFFER
+
+# Pad the frame size by 0x8 so when the xmm0 and xmm1 register store/restore happens
+# we can align to 16 and be guaranteed to not exceed the frame size
+.equ STACK_FUDGE_FACTOR, 0x8
+
+# Space to keep xmm0 and xmm1
+.equ SIZEOF_FP_ARG_SPILL, 0x10*2
+
+.equ OFFSETOF_FP_ARG_SPILL, SIZEOF_PROFILE_PLATFORM_SPECIFIC_DATA
```
Comment (noahfalk) [inline src/coreclr/src/vm/amd64/asmhelpers.S]: This needs to include STACK_FUDGE_FACTOR

PROFILE_PLATFORM_SPECIFIC_DATA is stored from [rsp, rsp+SIZEOF_PROFILE_SPECIFIC_DATA]. Xmm0 is stored at (rsp+OFFSETOF_FP_ARG_SPILL) & -16. Right now both of those constants are 08x*22 = 0xb0. If rsp was 8 we'd be storing Xmm0 starting at (0x8 + 0xb0) & -16 = 0xb0 but the last bytes of PROFILE_PLATFORM_SPECIFIC_DATA would be stored up until 0x8 + 0xb0 = 0xb8. The last 8 bytes of PROFILE_PLATFORM_SPECIFIC_DATA get overwritten when alignment slides xmm0 backwards 8 bytes.

**Review by noahfalk — APPROVED** [review]
I think you've still got one bug in there, but fix that and a few other minor comments and we are good to go.
